infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/CorfuReplicationClusterManagerAdapter.java

```diff
@@ -1,51 +1,59 @@
 package org.corfudb.infrastructure.logreplication.infrastructure.plugins;
 
-import lombok.Getter;
-
 import org.corfudb.infrastructure.logreplication.infrastructure.CorfuReplicationDiscoveryServiceAdapter;
 import org.corfudb.infrastructure.logreplication.proto.LogReplicationClusterInfo.TopologyConfigurationMsg;
 
-public abstract class CorfuReplicationClusterManagerAdapter {
-    @Getter
-    CorfuReplicationDiscoveryServiceAdapter corfuReplicationDiscoveryService;
+/**
+ * This is the interface for CorfuReplicationClusterManager.
+ * Implementation of this should have following members:
+ * 1. corfuReplicationDiscoveryService that is needed to notify the cluster configuration change.
+ * 2. localEndpoint that has the local node information.
+ *
+ */
+public interface CorfuReplicationClusterManagerAdapter {
 
-    @Getter
-    TopologyConfigurationMsg topologyConfig;
+    /**
+     *   Register the discovery service and start the cluster manager callback service.
+     *   Implementation of this function should call register(), register this ClusterManager at the upper
+     *   layer to get the callback for topology change.
+     */
+     void start(CorfuReplicationDiscoveryServiceAdapter corfuReplicationDiscoveryService);
 
-    public void connect(CorfuReplicationDiscoveryServiceAdapter corfuReplicationDiscoveryService) {
-        this.corfuReplicationDiscoveryService = corfuReplicationDiscoveryService;
-        start();
-    }
+     /**
+     * Set the localEndpoint
+     */
+    void setLocalEndpoint(String endpoint);
 
-    public synchronized TopologyConfigurationMsg fetchTopology() {
-        topologyConfig = queryTopologyConfig();
-        return topologyConfig;
-    }
+    // This is the currentTopology cached at the adapter.
+    public TopologyConfigurationMsg getTopologyConfig();
 
-    /**
-     * Will be called when the cluster change and a new configuration is sent over
-     *
-     * @param newTopologyConfigMsg
-     */
-    public synchronized void updateTopologyConfig(TopologyConfigurationMsg newTopologyConfigMsg) {
-        if (newTopologyConfigMsg.getTopologyConfigID() > topologyConfig.getTopologyConfigID()) {
-            topologyConfig = newTopologyConfigMsg;
-            corfuReplicationDiscoveryService.updateTopology(topologyConfig);
-        }
-    }
+    // This will talk to the real Cluster Manager and get the most current topology.
+    TopologyConfigurationMsg queryTopologyConfig();
```

Comment:
Is it possible that we can merge these two queryTopologyConfig functions? Maybe use a boolean parameter to indicate query from the local cache or not.